### PR TITLE
feature: Add metrics support (MAPCO-3337)

### PR DESCRIPTION
**Pull request by almog8k**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

Further  information:
 Creating Prometheus Metrics-Server on port 9185.
 Adding metrics suuport to the tiles-merger-service usign Prometheus-Net.
 Creating grafana Dash-Board for analyzing the Prometheus scraped data.

### CODE REVIEW

File: MergerLogic/Monitoring/Metrics/MetricsProvider.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        private readonly bool _enabled;
+        
+        public MetricsProvider(IConfigurationManager configuration)
+        {
+            var appInfo = Assembly.GetEntryAssembly()?.GetName();
+            string appName = appInfo?.Name ?? "MergerService";
+            this._registry = Prometheus.Metrics.DefaultRegistry;
+            this._registry.SetStaticLabels(new Dictionary<string, string>(){{"app", appName}});
+            this._buckets = new double[]{0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 15, 50, 250, 500};
+            this._enabled = configuration.GetConfiguration<bool>("METRICS", "enabled");
```
Comment (asafMasa) [inline MergerLogic/Monitoring/Metrics/MetricsProvider.cs:21]: check tht merics enabled false works and doesn't produce metrics

**Review by asafMasa — CHANGES_REQUESTED** [review]
Good Job 😎

File: MergerLogic/DataTypes/Data.cs [hunk 2]
```diff
@@ -270,13 +275,22 @@ public bool TileExists(Coord coord)
         public void UpdateTiles(IEnumerable<Tile> tiles)
         {
             this._logger.LogDebug($"[{MethodBase.GetCurrentMethod().Name}] update tiles started");
+            var updateTilesStopWatch = new Stopwatch();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/Data.cs:278]: because there is switch case inside the method

File: helm/templates/service.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  {{- $serviceName := include "service.fullname" . -}}
+  {{- if .Values.enabled -}}
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ $serviceName }}
+  labels:
+    app: {{ $chartName }}
+    component: {{ $chartName }}
+    environment: {{ include "gpkg-merger.environment" . }}
```
[the remark this answers is not in the document]
Comment (almog8k) [reply helm/templates/service.yaml:14]: you can find the definitions here: _helpers.tpl and the values defined at the chart.yaml

File: MergerCli/Properties/launchSettings.json [hunk 4]
```diff
@@ -1,13 +1,13 @@
 {
   "profiles": {
     "GPKG to S3": {
-      "workingDirectory": "",
+      "workingDirectory": "/home/almogk/Documents/raster-team/Tests/input/gpkgs",
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerCli/Properties/launchSettings.json]: done

File: MergerCli/appsettings.json [hunk 5]
```diff
@@ -12,7 +12,7 @@
   },
   "S3": {
     "url": "http://localhost:9000",
-    "bucket": "tiles",
+    "bucket": "raster",
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerCli/appsettings.json]: done

File: MergerLogic/DataTypes/Data.cs [hunk 6]
```diff
@@ -77,6 +80,8 @@ public abstract class Data<TUtilsType> : IData where TUtilsType : IDataUtils
         {
             string methodName = MethodBase.GetCurrentMethod().Name;
             this._container = container;
+            var metricsProvider = this._container.GetRequiredService<IMetricsProvider>();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/Data.cs]: done

File: helm/values-local.yaml [hunk 7]
```diff
@@ -0,0 +1,119 @@
+global:
```
[the remark this answers is not in the document]
Comment (almog8k) [reply helm/values-local.yaml]: done

File: helm/values.yaml [hunk 8]
```diff
@@ -99,8 +99,10 @@ env:
     ratio: 0.01
   metrics:
     enabled: true
-    url: http://localhost:4317/v1/metrics
-    interval: 5000
+    scrape: true
+    port: 9185
+    measurementBuckets: [0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 15, 50,
+      250, 500]
```
[the remark this answers is not in the document]
Comment (almog8k) [reply helm/values.yaml]: done

File: MergerService/Utils/TaskUtils.cs [hunk 9]
```diff
@@ -49,24 +52,33 @@ public class TaskUtils : ITaskUtils
         {
             using (this._activitySource.StartActivity("dequeue task"))
             {
+                Stopwatch taskInitializationStopwatch = new Stopwatch();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/Utils/TaskUtils.cs:55]: ?

File: MergerService/src/run.cs [hunk 10]
```diff
@@ -355,10 +388,14 @@ private void RunTask(MergeTask task, ITaskUtils taskUtils, string? managerCallba
...
                                 }
                             }
+                            totalWorkStopwatch.Stop();
+                            var seconds = totalWorkStopwatch.Elapsed.TotalSeconds;
+                            this._metricsProvider.TotalBatchWorkTimeHistogram()?.Observe(totalWorkStopwatch.Elapsed.TotalSeconds);
                         }
 
                         using (this._activitySource.StartActivity("saving tiles"))
                         {
+                            var updateTilesStopWatch = new Stopwatch();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/src/run.cs:398]: removed

File: MergerService/src/run.cs [hunk 11]
```diff
@@ -316,6 +337,8 @@ private void RunTask(MergeTask task, ITaskUtils taskUtils, string? managerCallba
                         // Go over the bounds of the current batch
                         using (this._activitySource.StartActivity($"[{methodName}] merging tiles"))
                         {
+                            var totalWorkStopwatch = new Stopwatch();
+                            totalWorkStopwatch.Start();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/src/run.cs:341]: done

File: MergerService/appsettings.Development.json [hunk 12]
```diff
@@ -34,9 +34,11 @@
     "ratio": 1
   },
   "METRICS": {
-    "enabled": false,
-    "url": "",
-    "interval": 3000
+    "enabled": true,
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/appsettings.Development.json]: done

File: MergerService/appsettings.json [hunk 13]
```diff
@@ -42,9 +42,11 @@
     "ratio": 1
   },
   "METRICS": {
-    "enabled": false,
-    "url": "",
-    "interval": 3000
+    "enabled": true,
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/appsettings.json]: done

File: MergerLogic/Utils/DataFactory.cs [hunk 14]
```diff
@@ -11,13 +12,15 @@ public class DataFactory : IDataFactory
         private readonly IPathUtils _pathUtils;
         private readonly IServiceProvider _container;
         private readonly ILogger _logger;
+        private readonly IMetricsProvider _metricsProvider;
 
-        public DataFactory(IConfigurationManager configuration, IPathUtils pathUtils, IServiceProvider container, ILogger<DataFactory> logger)
+        public DataFactory(IConfigurationManager configuration, IPathUtils pathUtils, IServiceProvider container, ILogger<DataFactory> logger, IMetricsProvider metricsProvider)
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/Utils/DataFactory.cs:17]: removed

File: MergerLogic/Monitoring/Metrics/MetricsProvider.cs [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+        private readonly double[] _buckets;
+        private readonly bool _enabled;
+        
+        public MetricsProvider(IConfigurationManager configuration)
+        {
+            var appInfo = Assembly.GetEntryAssembly()?.GetName();
+            string appName = appInfo?.Name ?? "MergerService";
+            this._registry = Prometheus.Metrics.DefaultRegistry;
+            this._registry.SetStaticLabels(new Dictionary<string, string>(){{"app", appName}});
+            this._buckets = new double[]{0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 15, 50, 250, 500};
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/Monitoring/Metrics/MetricsProvider.cs]: done

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 16]
```diff
@@ -1,105 +1,120 @@
...
-            baseImage.ColorSpace = ColorSpace.sRGB;
-
-            using (var srcPixels = baseImage.GetPixels())
-            {
-                /*
-                * If the scale is bigger than the tile size, then that means that the dst image needs to be represented as one pixel in the src.
-                * Calculation:
-                * scale = 2 ^ (dst_zoom - src_zoom)
-                * subTileShift = TILE_SIZE / scale
+            var upscaleStopWatch = new Stopwatch();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/ImageProcessing/TileScaler.cs:22]: done

File: MergerLogic/DataTypes/XYZ.cs [hunk 17]
```diff
@@ -1,11 +1,11 @@
 ﻿using MergerLogic.Batching;
+using MergerLogic.Monitoring.Metrics;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/XYZ.cs:2]: removed

File: MergerLogic/DataTypes/WMTS.cs [hunk 18]
```diff
@@ -1,11 +1,11 @@
 ﻿using MergerLogic.Batching;
+using MergerLogic.Monitoring.Metrics;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/WMTS.cs:2]: removed

File: MergerLogic/DataTypes/TMS.cs [hunk 19]
```diff
@@ -1,11 +1,11 @@
 ﻿using MergerLogic.Batching;
+using MergerLogic.Monitoring.Metrics;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/TMS.cs:2]: removed

File: MergerLogic/DataTypes/S3.cs [hunk 20]
```diff
@@ -2,6 +2,7 @@
 using Amazon.S3.Model;
 using MergerLogic.Batching;
 using MergerLogic.Clients;
+using MergerLogic.Monitoring.Metrics;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/S3.cs:5]: removed

File: MergerLogic/DataTypes/HttpDataSource.cs [hunk 21]
```diff
@@ -1,5 +1,6 @@
 ﻿using MergerLogic.Batching;
 using MergerLogic.Clients;
+using MergerLogic.Monitoring.Metrics;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/HttpDataSource.cs:3]: removed

File: MergerLogic/DataTypes/Fs.cs [hunk 22]
```diff
@@ -1,4 +1,5 @@
 using MergerLogic.Batching;
+using MergerLogic.Monitoring.Metrics;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/Fs.cs:2]: removed

File: MergerLogic/DataTypes/Gpkg.cs [hunk 23]
```diff
@@ -1,5 +1,6 @@
 using MergerLogic.Batching;
 using MergerLogic.Clients;
+using MergerLogic.Monitoring.Metrics;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/Gpkg.cs:3]: removed

File: MergerService/Utils/TaskUtils.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/Utils/TaskUtils.cs:55]: done

**Review by asafMasa — APPROVED** [review]

**Review by asafMasa — CHANGES_REQUESTED** [review]
add grafana dashboard assets

File: MergerLogic/MergerLogic.csproj [hunk 24]
```diff
@@ -17,11 +17,15 @@
     <PackageReference Include="Microsoft.Extensions.Http.Polly" Version="6.0.8" />
     <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="6.0.1" />
     <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="6.0.0" />
-    <PackageReference Include="OpenTelemetry" Version="1.3.0" />
-    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.3.0" />
-    <PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.3.0" />
-    <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.0.0-rc9.4" />
-    <PackageReference Include="OpenTelemetry.Instrumentation.Http" Version="1.0.0-rc9.4" />
+    <PackageReference Include="OpenTelemetry" Version="1.6.0-alpha.1" />
```
Comment (asafMasa) [inline MergerLogic/MergerLogic.csproj:20]: latest stable

File: MergerLogic/Monitoring/Metrics/MetricsProvider.cs [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+                "total_validation_time",
+                "Histogram of Total Validation time",
+                null,
+                this._buckets
+
+            );
+        }
+
+        private Histogram GetOrCreateHistogram(string metricName, string help, string[] labels = null, double[] buckets = null)
+        {
```
Comment (asafMasa) [inline MergerLogic/Monitoring/Metrics/MetricsProvider.cs:137]: try to fix this warnings

File: helm/templates/configmap.yaml [hunk 26]
```diff
@@ -29,10 +29,11 @@ data:
   TRACING__url: {{ $tracingUrl }}
   TRACING__ratio: {{ .Values.env.tracing.ratio | quote}}
   {{ end }}
+  METRICS__enabled: {{ .Values.env.metrics.enabled | quote }} 
   {{ if .Values.env.metrics.enabled }}
-  METRICS__enabled: 'true'
-  METRICS__url: {{ $metricsUrl }}
-  METRICS__interval: {{ .Values.env.metrics.interval | quote}}
+  METRICS__scrape: {{ .Values.env.metrics.scrape | quote}}
```
Comment (asafMasa) [inline helm/templates/configmap.yaml:34]: space

File: helm/templates/route.yaml [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+  {{- $chartName := include "gpkg-merger.name" . -}}
+  {{- $serviceName := include "service.fullname" . -}}
+  {{- $cloudProviderFlavor := include "gpkg-merger.cloudProviderFlavor" . -}}
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  name: {{ $releaseName }}-{{ $chartName }}-route
+  annotations:
+    haproxy.router.openshift.io/rewrite-target: /
+spec:
```
Comment (asafMasa) [inline helm/templates/route.yaml:11]: make sure rout is disabled by default

**Review by asafMasa — CHANGES_REQUESTED** [review]
Good Job!

File: MergerLogic/DataTypes/Data.cs [hunk 28]
```diff
@@ -45,6 +46,7 @@ public abstract class Data<TUtilsType> : IData where TUtilsType : IDataUtils
         protected delegate Tile? NullableTileConvertorFunction(Tile tile);
 
         protected IServiceProvider _container;
+        protected IMetricsProvider _metricsProvider;
```
Comment (asafMasa) [inline MergerLogic/DataTypes/Data.cs:49]: check if needs to be reverted

**Review by asafMasa — CHANGES_REQUESTED** [review]
more changes

File: helm/templates/configmap.yaml (hunk 26 above)
Comment (almog8k) [reply helm/templates/configmap.yaml:34]: done

File: MergerLogic/DataTypes/Data.cs (hunk 28 above)
Comment (almog8k) [reply MergerLogic/DataTypes/Data.cs]: removed

File: helm/templates/route.yaml (hunk 27 above)
Comment (almog8k) [reply helm/templates/route.yaml:15]: done

File: MergerLogic/Utils/DataFactory.cs [hunk 29]
```diff
@@ -12,6 +12,7 @@ public class DataFactory : IDataFactory
         private readonly IServiceProvider _container;
         private readonly ILogger _logger;
 
+
```
Comment (shimoncohen) [inline MergerLogic/Utils/DataFactory.cs:15]: Remove extra line.

File: MergerService/Utils/TaskUtils.cs [hunk 30]
```diff
@@ -53,18 +52,16 @@ public class TaskUtils : ITaskUtils
                 string url = new Uri(new Uri(_jobManagerUrl), relativeUri).ToString();
                 string? taskData = this._httpClient.PostData(url, null);
 
-                if (taskData is null)
-                {
-                    return null;
-                }
```
Comment (shimoncohen) [inline MergerService/Utils/TaskUtils.cs:59]: Why was this moved from here?

File: MergerService/src/run.cs [hunk 31]
```diff
@@ -204,6 +212,11 @@ public void Start()
                     }
                     finally
```
Comment (shimoncohen) [inline MergerService/src/run.cs:213]: Does `finally` execute after `continue` statement? I couldn't find an answer to this, we should test to make sure.

File: helm/values.yaml [hunk 32]
```diff
@@ -99,14 +99,23 @@ env:
...
-      memory: "128Mi"
+      cpu: "1"
+      memory: 4Gi
+
+route:
+  enabled: false
+  path: /
+  timeout:
+    enabled: false # defaults to 30s by openshift
+    duration: 60s # supported units (us, ms, s, m, h, d)
```
Comment (shimoncohen) [inline helm/values.yaml:121]: Add a new line at the end of the file.

File: helm/templates/service.yaml [hunk 33]
```diff
@@ -0,0 +1,34 @@
+
+{{- $releaseName := .Release.Name -}}
+  {{- $chartName := include "gpkg-merger.name" . -}}
+  {{- $cloudProviderFlavor := include "gpkg-merger.cloudProviderFlavor" . -}}
+  {{- $serviceName := include "service.fullname" . -}}
+  {{- if .Values.enabled -}}
```
Comment (shimoncohen) [inline helm/templates/service.yaml:6]: Remove indentation.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: Assets/Dashboards/MergerDashboard.json
[the remark this answers is not in the document]
Comment (almog8k) [reply Assets/Dashboards/MergerDashboard.json:1]: asaf requested to add the dashboards-json to the repository so we could use it inside out-net

File: MergerLogic/DataTypes/Data.cs [hunk 34]
```diff
@@ -53,14 +55,16 @@ public abstract class Data<TUtilsType> : IData where TUtilsType : IDataUtils
...
         public bool IsNew { get; set; }
         public bool IsOneXOne => this.Grid == Grid.OneXOne;
-        
+
         protected readonly int BatchSize;
         protected TUtilsType Utils;
         protected GetTileFromXYZFunction GetTile;
         protected readonly GetTileFromCoordFunction GetLastExistingTile;
         protected readonly IGeoUtils GeoUtils;
         protected readonly ILogger _logger;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/Data.cs:65]: done!

File: MergerLogic/DataTypes/Data.cs [hunk 35]
```diff
@@ -144,11 +149,13 @@ public abstract class Data<TUtilsType> : IData where TUtilsType : IDataUtils
             bool exists = this.Exists();
             if (!exists)
             {
-                if (this.IsBase) {
+                if (this.IsBase)
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/Data.cs:152]: ok

File: MergerLogic/DataTypes/Data.cs [hunk 36]
```diff
@@ -256,6 +267,8 @@ public bool TileExists(Coord coord)
 
         public Tile? GetCorrespondingTile(Coord coords, bool upscale)
         {
+            Stopwatch fetchTileStopWatch = new Stopwatch();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/Data.cs:270]: creating new Stop watch every time is more explicit and readable.
I changed to your suggestion of using StopWatch.StartNew().

File: MergerLogic/Utils/DataFactory.cs (hunk 29 above)
Comment (almog8k) [reply MergerLogic/Utils/DataFactory.cs:15]: done

File: MergerService/Utils/TaskUtils.cs (hunk 30 above)
Comment (almog8k) [reply MergerService/Utils/TaskUtils.cs:59]: combined to single line, but i will revert it so it could be more clear.

File: MergerService/src/run.cs (hunk 31 above)
Comment (almog8k) [reply MergerService/src/run.cs:213]: finally does not execute after continue, i changed the code and take the continue statement outside from the catch.
check if the logic is ok.

File: helm/values.yaml (hunk 32 above)
Comment (almog8k) [reply helm/values.yaml:121]: done

File: helm/templates/service.yaml (hunk 33 above)
Comment (almog8k) [reply helm/templates/service.yaml:6]: done

File: MergerLogic/Monitoring/Metrics/MetricsProvider.cs (hunk 25 above)
Comment (almog8k) [reply MergerLogic/Monitoring/Metrics/MetricsProvider.cs:137]: fixed

File: MergerLogic/DataTypes/Data.cs (hunk 36 above)
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/DataTypes/Data.cs:270]: In NET 7.0 there were some improvements in the Stopwatch class including the 'GetEllapesedTime()', but we are unfortunately we are using NET 6.0.
Regrading to StopWatch instances the memory usage is negligible.

[the remark this answers is not in the document]
Comment (shimoncohen) [reply MergerLogic/DataTypes/Data.cs]: Thank you very much, appreciate the effort!

File: MergerLogic/ImageProcessing/TileScaler.cs [hunk 37]
```diff
@@ -19,8 +19,7 @@ public TileScaler(IMetricsProvider metricsProvider)
 
         public MagickImage? Upscale(MagickImage baseImage, Tile baseTile, Coord targetCoords)
         {
-            var upscaleStopWatch = new Stopwatch();
-            upscaleStopWatch.Start();
+            var upscaleStopWatch = Stopwatch.StartNew();
```
Comment (shimoncohen) [inline MergerLogic/ImageProcessing/TileScaler.cs:22]: This variable only has this function as it's scope.
Please rename to something like `stopwatch`.

File: MergerCli/Program.cs [hunk 38]
```diff
@@ -93,8 +95,7 @@ private static void Main(string[] args)
                 bool validate = config.GetConfiguration<bool>("GENERAL", "validate");
                 for (int i = 1; i < sources.Count; i++)
                 {
-                    Stopwatch stopWatch = new Stopwatch();
-                    stopWatch.Start();
+                    Stopwatch stopWatch = Stopwatch.StartNew();
```
Comment (shimoncohen) [inline MergerCli/Program.cs:98]: Once you added another `Stopwatch` variable to this function there is a need for a clearer name for this one.

File: MergerLogic/DataTypes/Data.cs [hunk 39]
```diff
@@ -267,8 +265,7 @@ public bool TileExists(Coord coord)
 
         public Tile? GetCorrespondingTile(Coord coords, bool upscale)
         {
-            Stopwatch fetchTileStopWatch = new Stopwatch();
-            fetchTileStopWatch.Start();
+            Stopwatch fetchTileStopWatch = Stopwatch.StartNew();
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Data.cs:268]: This variable only has this function as it's scope.
Please rename to something like `stopwatch`.

File: MergerLogic/DataTypes/Data.cs [hunk 40]
```diff
@@ -278,14 +275,13 @@ public bool TileExists(Coord coord)
...
-            this._metricsProvider.TotalFetchTimePerTileHistogram()?.Observe(fetchTileStopWatch.Elapsed.TotalMilliseconds);
+            this._metricsProvider.TotalFetchTimePerTileHistogram(fetchTileStopWatch.Elapsed.TotalMilliseconds);
             return correspondingTile;
         }
 
         public void UpdateTiles(IEnumerable<Tile> tiles)
         {
-            var batchUploadStopWatch = new Stopwatch();
-            batchUploadStopWatch.Start();
+            var batchUploadStopWatch = Stopwatch.StartNew();
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Data.cs:284]: This variable only has this function as it's scope.
Please rename to something like `stopwatch`.

File: MergerLogic/Extensions/ServiceCollectionExtensions.cs [hunk 41]
```diff
@@ -178,25 +180,25 @@ public static IServiceCollection RegisterOpenTelemetry(this IServiceCollection c
...
+                int metricsPort = _config.GetConfiguration<int>("METRICS", "port");
                 MetricServer metricsServer = new MetricServer(port: metricsPort);
                 metricsServer.Start();
             }
-            
+
             return collection;
         }
-        
+
```
Comment (shimoncohen) [inline MergerLogic/Extensions/ServiceCollectionExtensions.cs:202]: Please remove extra lines.

File: helm/values.yaml [hunk 42]
```diff
@@ -118,4 +118,5 @@ route:
   path: /
   timeout:
     enabled: false # defaults to 30s by openshift
-    duration: 60s # supported units (us, ms, s, m, h, d)
\ No newline at end of file
+    duration: 60s # supported units (us, ms, s, m, h, d)
+
```
Comment (shimoncohen) [inline helm/values.yaml:122]: There seems to be some weird character at the last line here.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerService/src/run.cs (hunk 31 above)
Comment (shimoncohen) [reply MergerService/src/run.cs:210]: From what I can see this looks good, please be sure to check that it works in all cases.

File: MergerLogic/Extensions/ServiceCollectionExtensions.cs [hunk 43]
```diff
@@ -145,9 +148,9 @@ public static IServiceCollection RegisterOpenTelemetry(this IServiceCollection c
                         //console exporter has hardcoded multiline log format so it cant be used
                         //options.AddConsoleExporter();
                         //disable lgtm not disposed alert as it is global singleton that is used for the entire app life
-                        options.AddProcessor(
-                            new SimpleLogRecordExportProcessor(new OpenTelemetryFormattedConsoleExporter(new ConsoleExporterOptions()))); //lgtm [cs/local-not-disposed]
                         options.SetResourceBuilder(resourceBuilder);
+                        options.AddProcessor(
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/Extensions/ServiceCollectionExtensions.cs:154]: revert as it's used to be

File: MergerService/src/run.cs (hunk 31 above)
Comment (almog8k) [reply MergerService/src/run.cs:210]: checked

File: MergerService/src/run.cs [hunk 44]
```diff
@@ -94,8 +94,8 @@ private List<IData> BuildDataList(Source[] paths, int batchSize)
         {
             using (this._activitySource.StartActivity("sources parsing"))
             {
-                var buildSourcesListStopWatch = new Stopwatch();
-                buildSourcesListStopWatch.Start();
+                var buildSourcesListStopWatch = Stopwatch.StartNew();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/src/run.cs:97]: done

File: MergerLogic/Monitoring/Metrics/MetricsProvider.cs [hunk 45]
```diff
@@ -21,146 +39,168 @@ public MetricsProvider(IConfigurationManager configurationManager)
...
+            this.ObserveHistogram
             (
-                "task_execution_time",
+                MetricName.TaskExecutionTimeHistogram,
                 "Histogram of task execution times in seconds",
+                value,
                 new string[] { "task_type", "success" },
-                this._buckets
-            );
+                labelValues
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/Monitoring/Metrics/MetricsProvider.cs:50]: done

File: MergerLogic/Monitoring/Metrics/MetricsProvider.cs [hunk 46]
```diff
@@ -21,146 +39,168 @@ public MetricsProvider(IConfigurationManager configurationManager)
...
+                value,
                 null,
-                this._buckets
-
+                labelValues
             );
         }
 
-        private Histogram? GetOrCreateHistogram(string metricName, string help, string[]? labels = null, double[]? buckets = null)
+        private void ObserveHistogram(MetricName metricName, string help, double value, string[]? labels = null, string[]? labelsValues = null)
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/Monitoring/Metrics/MetricsProvider.cs:163]: done

File: MergerLogic/Monitoring/Metrics/MetricsProvider.cs [hunk 47]
```diff
@@ -21,146 +39,168 @@ public MetricsProvider(IConfigurationManager configurationManager)
...
+            if (labelsValues != null)
+            {
+                histogram.WithLabels(labelsValues);
+            }
+            histogram.Observe(value);
         }
 
-        private Gauge? GetOrCreateGauge(string metricName, string help, string[]? labels = null)
+
+        private void GetOrCreateGauge(MetricName metricName, string help, double value, string[]? labels = null, string[]? labelsValues = null)
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerLogic/Monitoring/Metrics/MetricsProvider.cs:185]: done

File: MergerService/src/run.cs [hunk 48]
```diff
@@ -186,14 +186,13 @@ public void Start()
                     string log = managerCallbackUrl == null ? "managerCallbackUrl not provided as job parameter" : $"managerCallback url: {managerCallbackUrl}";
                     this._logger.LogDebug($"[{methodName}]{log}");
                     var totalTaskStopWatch = Stopwatch.StartNew();
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/src/run.cs:188]: change all to Stopwatch

File: MergerLogic/ImageProcessing/TileScaler.cs (hunk 37 above)
Comment (almog8k) [reply MergerLogic/ImageProcessing/TileScaler.cs:22]: done

File: MergerCli/Program.cs (hunk 38 above)
Comment (almog8k) [reply MergerCli/Program.cs:98]: changed to mergeRunTimeStopwatch

File: MergerLogic/DataTypes/Data.cs (hunk 40 above)
Comment (almog8k) [reply MergerLogic/DataTypes/Data.cs:284]: done

File: MergerService/src/run.cs [hunk 49]
```diff
@@ -387,16 +382,14 @@ private void RunTask(MergeTask task, ITaskUtils taskUtils, string? managerCallba
...
-                            this._metricsProvider
-                                .BatchWorkTimeHistogram()?
-                                .Observe(batchWorkTimeStopWatch.Elapsed.TotalSeconds);
+                            this._metricsProvider.BatchWorkTimeHistogram(batchWorkTimeStopWatch.Elapsed.TotalSeconds); ;
                         }
 
                         using (this._activitySource.StartActivity("saving tiles"))
                         {
-                          
+
```
[the remark this answers is not in the document]
Comment (almog8k) [reply MergerService/src/run.cs:390]: done

File: MergerLogic/Monitoring/Metrics/MetricsProvider.cs [hunk 50]
```diff
@@ -1,10 +1,28 @@
 
+using Amazon.Runtime;
 using MergerLogic.Utils;
 using Prometheus;
 using System.Reflection;
+using System.Runtime.Serialization;
 
 namespace MergerLogic.Monitoring.Metrics
 {
+    public enum MetricName
```
Comment (shimoncohen) [inline MergerLogic/Monitoring/Metrics/MetricsProvider.cs:10]: Please make this `private`.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

**Review by shimoncohen — APPROVED** [review]

**Review by asafMasa — APPROVED** [review]
